### PR TITLE
fix(fd): start at beginning of challenge period

### PR DESCRIPTION
**Pull request by smartcontracts**
**Description**
Updates the fault-detector to use a smarter starting height by default.
Specifically, the fault-detector will now start at the first batch that
has not yet crossed the challenge period, since the general trust
assumption is that invalid batches will not pass the challenge period.
Significantly reduces verification time, especially since the fault
detector is stateless and does not keep track of the highest verified
batch.

### CODE REVIEW

File: packages/fault-detector/src/service.ts [hunk 1]
```diff
@@ -41,7 +42,7 @@ export class FaultDetector extends BaseServiceV2<Options, Metrics, State> {
         },
         startBatchIndex: {
           validator: validators.num,
-          default: 0,
+          default: -1,
```
Comment (smartcontracts) [inline packages/fault-detector/src/service.ts:45]: `BaseServiceV2` doesn't cleanly handle "undefined" defaults but -1 is good enough here.

File: packages/fault-detector/src/service.ts [hunk 2]
```diff
@@ -67,19 +68,56 @@ export class FaultDetector extends BaseServiceV2<Options, Metrics, State> {
...
+      let end = (await this.state.scc.getTotalBatches()).toNumber()
+      while (start !== end) {
+        const mid = Math.ceil((start + end) / 2)
+        const filter = this.state.scc.filters.StateBatchAppended(mid)
+        const events = await this.state.scc.queryFilter(filter)
+        const block = await this.options.l1RpcProvider.getBlock(
+          events[0].blockNumber
+        )
+
+        if (block.timestamp > latestBlock.timestamp - fpw) {
```
Comment (smartcontracts) [inline packages/fault-detector/src/service.ts:102]: Will converge on index 0 if the chain is less than `fpw` seconds old.

**Comment by tynes** [comment]
How exactly are you testing this?

**Comment by smartcontracts** [comment]
> How exactly are you testing this?

Added tests

File: packages/fault-detector/package.json [hunk 3]
```diff
@@ -32,13 +33,22 @@
...
+    "chai-as-promised": "^7.1.1",
     "dateformat": "^4.5.1",
+    "ethereum-waffle": "^3.4.4",
     "ethers": "^5.6.8",
+    "hardhat": "^2.9.6",
+    "lodash": "^4.17.21",
     "ts-node": "^10.7.0"
   },
   "dependencies": {
+    "@defi-wonderland/smock": "^2.0.7",
```
Comment (tynes) [inline packages/fault-detector/package.json:49]: Should this be a devdep? Not that big of a deal

**Review by tynes — APPROVED** [review]

**Review by mslipper — APPROVED** [review]

**Pull request merged**